### PR TITLE
mavlink_messages: Subscribe to all distance_sensor instances

### PR DESCRIPTION
**Pull request by mortenfyhn**
**Describe problem solved by this pull request**
If you have more than one distance sensor, they might publish to different uorb distance sensor topic instances. The module that picks up these messages and forwards them to mavlink only subscribes to the first instance, so you won't get all the data. I also believe that which sensor gets which uorb instance id is non-deterministic, so it's random at startup which sensor you'll get data from on mavlink.

**Describe your solution**
Subscribe to all distance_sensor uorb instances, and send everything on mavlink. This is how for instance `MavlinkStreamBatteryStatus` is implemented.

**Describe possible alternatives**
Another option suggested is to parametrize the distance sensor mavlink streamer class with a template, as discussed with @JacobCrabill on the slack:

I'm interested in opinions on this! My implementation is kind of "the bare minimum". A different solution might be better, but I think the functionality (getting mavlink data from multiple distance sensors) is useful.

a

### CODE REVIEW

File: src/modules/mavlink/mavlink_messages.cpp [hunk 1]
```diff
@@ -4503,64 +4503,74 @@ class MavlinkStreamDistanceSensor : public MavlinkStream
...
 	unsigned get_size() override
 	{
-		return _distance_sensor_sub->is_published() ? (MAVLINK_MSG_ID_DISTANCE_SENSOR_LEN + MAVLINK_NUM_NON_PAYLOAD_BYTES) : 0;
+		return MAVLINK_MSG_ID_DISTANCE_SENSOR_LEN + MAVLINK_NUM_NON_PAYLOAD_BYTES;
 	}
 
 private:
-	MavlinkOrbSubscription *_distance_sensor_sub;
-	uint64_t _dist_sensor_time;
+	MavlinkOrbSubscription *_distance_sensor_sub[ORB_MULTI_MAX_INSTANCES] {};
```
Comment (JacobCrabill) [inline src/modules/mavlink/mavlink_messages.cpp]: Ah, yeah.  This is a little simpler than my suggestion of using templates for all the instances you want.

Comment (mortenfyhn) [reply src/modules/mavlink/mavlink_messages.cpp]: Yeah, I just did the same as `MavlinkStreamBatteryStatus` does. Tested on a hardware and I seem to be getting data fine.

**Comment by dagar** [comment]
It seems like there should be a better way to do this (separate mavlink streams per instance perhaps), but this should be fine if `get_size()` can be fixed.

File: src/modules/mavlink/mavlink_messages.cpp [hunk 2]
```diff
@@ -4503,64 +4503,74 @@ class MavlinkStreamDistanceSensor : public MavlinkStream
 
 	unsigned get_size() override
 	{
-		return _distance_sensor_sub->is_published() ? (MAVLINK_MSG_ID_DISTANCE_SENSOR_LEN + MAVLINK_NUM_NON_PAYLOAD_BYTES) : 0;
+		return MAVLINK_MSG_ID_DISTANCE_SENSOR_LEN + MAVLINK_NUM_NON_PAYLOAD_BYTES;
```
[the remark this answers is not in the document]
Comment (mortenfyhn) [reply src/modules/mavlink/mavlink_messages.cpp]: Loop over instances and sum up bytes for the instances where `is_published` is true?

**Comment by mortenfyhn** [comment]
Would something like this work?

```cpp
unsigned get_size() override
{
	int size = 0;

	for (int i = 0; i < ORB_MULTI_MAX_INSTANCES; i++) {
		if (_distance_sensor_sub[i] && _distance_sensor_sub[i]->is_published()) {
			size += MAVLINK_MSG_ID_DISTANCE_SENSOR_LEN + MAVLINK_NUM_NON_PAYLOAD_BYTES;
		}
	}

	return size;
}
```

The class I based this on, `MavlinkStreamBatteryStatus` seems to just `return MAVLINK_MSG_ID_BATTERY_STATUS_LEN + MAVLINK_NUM_NON_PAYLOAD_BYTES;`, but that doesn't mean it's right, of course.

**Comment by julianoes** [comment]
Ping @dagar

**Comment by mortenfyhn** [comment]
Closing this in favour of #15698.

**Pull request closed without merging**